### ISSUE TITLE
Ability to grab track metadata (tempo change messages?)

### ISSUE THREAD

**Issue by kavika13** [schellingb/TinySoundFont#28]
I'm converting an old DirectMusic application to use TML. I need to seek/loop to a specific part of the song, but the application I am using has the time stamps encoded in `MUSIC_TIME` (which is relative to tempo). In order to determine the actual milliseconds offset within the song, I need to seek from the beginning and see all tempo changes.

If the midi header has tempo data, it would also be useful to have that available, in case no change tempo messages are in the stream.

**Comment by schellingb** [comment]
Because tml consists only of messages after loading it would either have to create a fake message which contains the divison (ticks per beat) of the MIDI header or extend tml_load with an optional second parameter that gets filled with that value. Other than that there is no tempo defined in the header. It is defaulted to 480000 for MIDI until a SET_TEMPO arrives.
Leaving the SET_TEMPO messages in the accessible messages would be possible but due to the layout of tml_message it would be kinda weird/confusing because the 3 byte integer tempo value ends up in `channel`, `key` and `velocity`. Adding even more unions would make this less weird in one way but more confusing especially for projects that don't need the tempo info.
Maybe leave it weird and have a `int tml_get_tempo_value(tml_message* set_tempo_message)` function that returns the 3 bytes value properly would be a good solution.

**Comment by kavika13** [comment]
> Speaking of that default tempo, the value you mentioned (and in the code) seems wrong - Shouldn't it be 500000? When I googled, I saw that fluid synth has that 480000 value, but as far as I can tell, it's a bug there too (?).
> 
> I have a midi file with division = 240, at 120bpm, and the first note is at 1.85 seconds. The ticks value for that is 888.
> 
> 888 ticks * 480000 [µs/quarter note] / 240 [ticks/quarter note] / 1000 [µs/ms] = 1776 ms
> Having looked at this in a separate midi debugger, the 1776 ms value is definitely wrong.
> 
> The midi file also has a SET_TEMPO at 0 delta time, with a value of 500000 (120bpm), so:
> 888 ticks * 500000 [µs/quarter note] / 240 [ticks/quarter note] / 1000 [µs/ms] = 1850 ms
> 
> Should I file another bug for this?

I tried commenting out the `SET_TEMPO` reaction code, and tested the song length vs a different midi player. When I have the constant set to `500000`, I get a track length of 7.95s, which matches my other midi player. When I have the constant set to `480000`, as it currently is in the code, I get a track length of 7.632s.

So yeah, I'm thinking bug.

**Comment by schellingb** [comment]
I just committed 5506592
>  Leave MIDI messages of type SET_TEMPO accessible in the linked list
> Add tml_get_tempo_value function to read the tempo value as int

So the TML_SET_TEMPO enum entry is now in the public header and such messages are in the linked list like any other channel based message.

A cleaner way of reading metadata would be nice but for now this is a minimal change. The SET_TEMPO messages were already allocated and in memory anyway, they just weren't accessible by the linked list.

A decade or so ago I showed embedded lyrics alongside MIDI playback in some game I made. I always wanted to support text metadata in tml but it either gets kinda hacky or it doesn't feel 'Tiny' anymore. Maybe one day (-:
